lib/web/desktop_playback.go

```diff
@@ -0,0 +1,237 @@
+func (pp *playbackPlayer) close(cancel context.CancelFunc) {
+	pp.closeOnce.Do(func() {
+		pp.mu.Lock()
+		defer pp.mu.Unlock()
+
+		err := pp.ws.Close()
+		if err != nil {
+			pp.log.WithError(err).Errorf("websocket.Close() failed")
+		}
+
+		pp.playState = playStateFinished
+		pp.cond.Broadcast()
+		cancel()
+	})
+}
+
+// receiveActions handles logic for recieving playbackAction jsons
+// over the websocket and modifying playbackPlayer's state accordingly.
+func (pp *playbackPlayer) receiveActions(cancel context.CancelFunc) {
+	defer pp.log.Debug("playbackPlayer.ReceiveActions returned")
+	defer pp.close(cancel)
+
+	for {
+		action := playbackAction{}
+		// Hangs until there is data to be received, or until an error.
+		err := websocket.JSON.Receive(pp.ws, &action)
+		if err != nil {
+			// We expect net.ErrClosed if the websocket is closed by another
+			// goroutine and io.EOF if the websocket is closed by the browser
+			// while websocket.JSON.Receive() is hanging.
+			if !(errors.Is(err, net.ErrClosed) || errors.Is(err, io.EOF)) {
+				pp.log.WithError(err).Error("error reading from websocket")
+			}
+			return
+		}
+		pp.log.Debugf("recieved playback action: %+v", action)
+		if action.Action == actionPlayPause {
+			pp.togglePlaying()
+		} else {
+			pp.log.Errorf("received unknown action: %v", action.Action)
+			return
+		}
+	}
+}
+
+// streamSessionEvents streams the session's events as playback events over the websocket.
+func (pp *playbackPlayer) streamSessionEvents(ctx context.Context, cancel context.CancelFunc) {
+	defer pp.log.Debug("playbackPlayer.StreamSessionEvents returned")
+	defer pp.close(cancel)
+
+	var lastDelay int64
+	eventsC, errC := pp.clt.StreamSessionEvents(ctx, session.ID(pp.sID), 0)
+	for {
+		pp.waitWhilePaused()
+
+		select {
+		case err := <-errC:
+			if !errors.Is(err, context.Canceled) {
+				pp.log.WithError(err).Errorf("streaming session %v", pp.sID)
+			}
+			return
+		case evt := <-eventsC:
+			if evt == nil {
+				pp.log.Debug("reached end of playback")
+				// deferred ps.Close() will set ps.playState = playStateFinished for us
+				return
+			}
+			switch e := evt.(type) {
+			case *apievents.DesktopRecording:
+				if e.DelayMilliseconds > lastDelay {
+					time.Sleep(time.Duration(e.DelayMilliseconds-lastDelay) * time.Millisecond)
+					lastDelay = e.DelayMilliseconds
+				}
+				if _, err := pp.ws.Write(e.Message); err != nil {
+					// We expect net.ErrClosed to arise when another goroutine returns before
+					// this one or the browser window is closed, both of which cause the websocket to close.
+					if !errors.Is(err, net.ErrClosed) {
+						pp.log.WithError(err).Error("failed to write TDP message over websocket")
+					}
+					return
+				}
+			default:
+				pp.log.Warnf("session %v contains unexpected event type %T", pp.sID, evt)
+			}
+		}
+	}
+}
+
+// wait waits until the playbackPlayer's context is cancelled.
+func (pp *playbackPlayer) wait(ctx context.Context, cancel context.CancelFunc) {
```

Comment:
I would remove the wait func entirely and inline its logic at the end of `Play()`.

Since wait now takes an arbitrary context, there's no guarantee that the caller passes the same context used by the player, making this easy to mis-use.